### ISSUE TITLE
Grails 3.2.1 classloader issues with domain classes

### ISSUE THREAD

**Issue by gregopet** [grails/grails-core#10244]
I upgraded a Grails 3.0.9 application to Grails 3.2.1 (running on Oracle's Java 8) and have started getting interesting issues. I have a controller in which I fetch a List of objects (GORM + Hibernate 4): `List<Interstitial> interstitials = Interstitial.list()`

Later I want to extract some properties from them with `def interstitialConfig = interstitials?.collect { Interstitial instance -> ...}` and I get the following error:

```
MissingMethodException occurred when processing request: [GET] /api/ios/1/foo.json
No signature of method: foo.bar.MobileApplicationController$_foo_closure8.doCall() is applicable for argument types: (foo.bar.Interstitial) values: [foo.bar.Interstitial : 1]
Possible solutions: doCall(foo.bar.Interstitial), findAll(), findAll(), isCase(java.lang.Object), isCase(java.lang.Object)
The following classes appear as argument class and as parameter class, but are defined by different class loader:
foo.bar.Interstitial (defined by 'org.springframework.boot.loader.LaunchedURLClassLoader@77459877' and 'grails.boot.config.tools.ClassPathScanner$ApplicationRelativeClassLoader@33f676f6')
If one of the method suggestions matches the method you wanted to call, 
then check your class loader setup..
```

The stacktrace ends at the line containing the `collect` so it didn't seem relevant, but I can paste it if it helps. The bug also only appears on a staging/production build, not in development - we're deploying using self contained JARs.

We're using quite a few plugins, prime suspects for this could be Spring Security, Joda Time + Jadira mappings, PostgreSQL extensions, Build Test Data, Quartz. I can list them all if it helps. I can try to prepare an application to reproduce the bug, but I can't get around to it before this evening at the earliest.

**Comment by graemerocher** [comment]
Can probably be worked around by overriding `limitScanningToApplication()` in your `Application` class and returning `false`

**Comment by sagrawal14** [comment]
Let me check this workaround

**Comment by sagrawal14** [comment]
Graeme, this workaround seems to be working fine for me. Thanks for the support.

**Comment by gregopet** [comment]
Workaround works for me, too, many thanks!

**Comment by graemerocher** [comment]
@gregopet @sagrawal14 could you also try and update `gradle.properties` and use 

```
grailsVersion=3.2.2.BUILD-SNAPSHOT
```

To verify that the problem is fixed.

**Comment by sagrawal14** [comment]
Sure, let me try that as well.

**Comment by gregopet** [comment]
Works!

Commented out `limitScanningToApplication` override, ran `gradle clean` and deployed to staging, no errors.

**Comment by graemerocher** [comment]
Thanks for confirming

**Comment by sagrawal14** [comment]
Yes, working!

**Comment by runwolf** [comment]
We had an issue deploying to tomcat, using grails 3.2.1 if we have:

grailsVersion=3.2.1 in the gradle.prop, we are getting error like:

`ERROR org.grails.web.errors.GrailsExceptionResolver - GroovyCastException occurred when processing request: [GET] /LeadAgentBeta/web/getImage/33
Cannot cast object 'com.la.ImageFile : 33' with class 'com.la.ImageFile' to class 'com.la.ImageFile'. Stacktrace follows:
org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object 'com.la.ImageFile : 33' with class 'com.la.ImageFile' to class 'com.la.ImageFile'`

If we change it to 3.2.0 it's ok
and if we change it to grailsVersion=3.2.2.BUILD-SNAPSHOT it's ok too
if we run it in dev within grails it's ok.

so I am guessing it's fix in 3.2.2 :-)

**Comment by sagrawal14** [comment]
Yes @runwolf they fixed it in 3.2.2 but that is not yet released so just change `grailsVersion=3.2.2.BUILD-SNAPSHOT` like you are doing
